### ISSUE TITLE
EdgeHub loses direct method connection in 1.0.7

### ISSUE THREAD

**Issue by yfakariya** [Azure/iotedge#1466]
I faced lost of direct method connection between IoTHub and a edge module via AMQP. It looks another pattern of #646.

## Expected Behavior

* The edge hub never loses it connection due to `NullReferenceException`.
* The edge hub reconnect to the IoTHub successfully.

## Current Behavior

* The edge hub throws `NullReferenceException` in `DisposeMessageAsync()` call.
* The edge hub reports that it reconnect to the IoTHub successfully, but the IoTHub does not recognize it.

## Steps to Reproduce

I've not reproduced this issue yet.

### Output of `iotedge check`

I cannot run the command on the machine, sorry.

### Device (Host) Operating System

Debian 9

### Architecture

x64

### Container Operating System

Linux

#### iotedged

1.0.7.1

#### Edge Agent

1.0.7

#### Edge Hub

1.0.7

#### Docker

(I wrote from our setup log)

* moby-engine: 3.0.6
* moby-cli: 3.0.6

## Logs

<details>
<summary>edge-hub logs</summary>

```
2019-07-17 20:45:31.620 +00:00 [INF] - Closing receiver in cloud proxy b6e883e1-ae16-4352-bfe1-4e75119748e9 for MY_DEVICE_ID/TROUBLED_MODULE_ID
2019-07-17 20:45:31.623 +00:00 [INF] - Closed cloud proxy b6e883e1-ae16-4352-bfe1-4e75119748e9 for MY_DEVICE_ID/TROUBLED_MODULE_ID
2019-07-17 20:45:31.634 +00:00 [INF] - Retrying cloud proxy operation SendFeedbackMessageAsync for MY_DEVICE_ID/TROUBLED_MODULE_ID.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Azure.Devices.Client.Transport.Amqp.AmqpTransportHandler.DisposeMessageAsync(String lockToken, Outcome outcome, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.<>c__DisplayClass22_0.<<ExecuteWithErrorHandlingAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<>c__DisplayClass26_0.<<CompleteAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.CompleteAsync(String lockToken, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.InternalClient.CompleteAsync(String lockToken)
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.<>c__DisplayClass28_0.<<InvokeFunc>b__0>d.MoveNext() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ConnectivityAwareClient.cs:line 174
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.InvokeFunc[T](Func`1 func, String operation, Boolean useForConnectivityCheck) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ConnectivityAwareClient.cs:line 134
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudProxy.SendFeedbackMessageAsync(String messageId, FeedbackStatus feedbackStatus) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudProxy.cs:line 192
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudProxy.SendFeedbackMessageAsync(String messageId, FeedbackStatus feedbackStatus) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudProxy.cs:line 211
   at Microsoft.Azure.Devices.Edge.Hub.Core.Cloud.RetryingCloudProxy.<>c__DisplayClass22_0.<<ExecuteOperation>b__0>d.MoveNext() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/cloud/RetryingCloudProxy.cs:line 58
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Edge.Hub.Core.Cloud.RetryingCloudProxy.ExecuteOperation[T](Func`2 func, String operation) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/cloud/RetryingCloudProxy.cs:line 71
2019-07-17 20:45:31.809 +00:00 [INF] - Attempting to connect to IoT Hub for client MY_DEVICE_ID/TROUBLED_MODULE_ID via AMQP...
2019-07-17 20:45:32.407 +00:00 [INF] - Cloud connection for MY_DEVICE_ID/TROUBLED_MODULE_ID is True
2019-07-17 20:45:32.408 +00:00 [INF] - Connection status for MY_DEVICE_ID/TROUBLED_MODULE_ID changed to ConnectionEstablished
2019-07-17 20:45:32.408 +00:00 [INF] - Created cloud proxy for client MY_DEVICE_ID/TROUBLED_MODULE_ID via AMQP, with client operation timeout 20 seconds.
2019-07-17 20:45:32.411 +00:00 [INF] - Initialized cloud proxy 1af2641d-1907-4828-bece-a3fd4f251d09 for MY_DEVICE_ID/TROUBLED_MODULE_ID
2019-07-17 20:45:32.411 +00:00 [INF] - Created cloud connection for client MY_DEVICE_ID/TROUBLED_MODULE_ID
2019-07-17 20:45:32.574 +00:00 [INF] - Closing receiver in cloud proxy 1af2641d-1907-4828-bece-a3fd4f251d09 for MY_DEVICE_ID/TROUBLED_MODULE_ID
2019-07-17 20:45:32.575 +00:00 [INF] - Closed cloud proxy 1af2641d-1907-4828-bece-a3fd4f251d09 for MY_DEVICE_ID/TROUBLED_MODULE_ID
2019-07-17 20:45:32.625 +00:00 [WRN] - Error disposing message in ModuleMessages link for MY_DEVICE_ID/TROUBLED_MODULE_ID
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Azure.Devices.Client.Transport.Amqp.AmqpTransportHandler.DisposeMessageAsync(String lockToken, Outcome outcome, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.<>c__DisplayClass22_0.<<ExecuteWithErrorHandlingAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<>c__DisplayClass26_0.<<CompleteAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.CompleteAsync(String lockToken, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.InternalClient.CompleteAsync(String lockToken)
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.<>c__DisplayClass28_0.<<InvokeFunc>b__0>d.MoveNext() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ConnectivityAwareClient.cs:line 174
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.InvokeFunc[T](Func`1 func, String operation, Boolean useForConnectivityCheck) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ConnectivityAwareClient.cs:line 134
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudProxy.SendFeedbackMessageAsync(String messageId, FeedbackStatus feedbackStatus) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudProxy.cs:line 192
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudProxy.SendFeedbackMessageAsync(String messageId, FeedbackStatus feedbackStatus) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudProxy.cs:line 211
   at Microsoft.Azure.Devices.Edge.Hub.Core.Cloud.RetryingCloudProxy.<>c__DisplayClass22_0.<<ExecuteOperation>b__0>d.MoveNext() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/cloud/RetryingCloudProxy.cs:line 58
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Edge.Hub.Core.Cloud.RetryingCloudProxy.ExecuteOperation[T](Func`2 func, String operation) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/cloud/RetryingCloudProxy.cs:line 71
   at Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler.ProcessMessageFeedbackAsync(String messageId, FeedbackStatus feedbackStatus) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs:line 109
   at Microsoft.Azure.Devices.Edge.Hub.Amqp.LinkHandlers.SendingLinkHandler.DisposeMessageAsync(Delivery delivery) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Amqp/linkhandlers/SendingLinkHandler.cs:line 132
2019-07-17 20:45:35.383 +00:00 [INF] - Attempting to connect to IoT Hub for client MY_DEVICE_ID/TROUBLED_MODULE_ID via AMQP...
2019-07-17 20:45:35.992 +00:00 [INF] - Cloud connection for MY_DEVICE_ID/TROUBLED_MODULE_ID is True
2019-07-17 20:45:35.992 +00:00 [INF] - Connection status for MY_DEVICE_ID/TROUBLED_MODULE_ID changed to ConnectionEstablished
2019-07-17 20:45:35.995 +00:00 [INF] - Created cloud proxy for client MY_DEVICE_ID/TROUBLED_MODULE_ID via AMQP, with client operation timeout 20 seconds.
2019-07-17 20:45:35.996 +00:00 [INF] - Initialized cloud proxy 0e94606f-22b3-443e-861b-d42d9008067c for MY_DEVICE_ID/TROUBLED_MODULE_ID
2019-07-17 20:45:35.996 +00:00 [INF] - Created cloud connection for client MY_DEVICE_ID/TROUBLED_MODULE_ID
```
</details>

## Additional Information

* I noticed that the IoTHub had started to return HTTP 504 (gateway timeout) continuously from around 2019-07-17T20:45:30.
* I finally rebooted the edge machine and then it worked fine.
* I looked the edge hub's log, we also noticed that `NullReferenceException` had been occurred in the edge hub's `CloudProxy`, it tried to reconnect the IoTHub soon, and then the edge hub recognized that the cloud connection recovered successfully.
    * Module twins' communications were not affected.
* I took a support ticket to ask for investigation this problem, they said that there are no problems from the IoTHub perspective, and AMQP logical connection was not recovered in the IoTHub side.

**Comment by varunpuranik** [comment]
Hi @yfakariya, we have made a number of connectivity fixes in the IoT Edge runtime that are part of the latest release - 1.0.8. 
Can you please try to update to the 1.0.8 bits and see if you still get this issue? 

Please note that the Device SDK could throw the null ref exception, but the EdgeHub should be able to handle it.

**Comment by arsing** [comment]
1.0.8 is released.

**Comment by yfakariya** [comment]
@varunpuranik Unfortunately, it seems to be not fixed yet in 1.0.8. I open the new issue as #1541 because this issue's title says that this is 1.0.7 issue and closed in last month.
